Make updates of many un-Scoped factories state safer

ettavolt opened the pull request:
Also less locking when using Scope.

## Impact

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes expected existing functionality)
- [ ] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Description

Follow up #2018.

Migrating 11 databases with four threads rarely led to an NPE due to concurrency in ReflectionSerializer.
I've decided to not stop with just that but also look at all static modifiable fields and make access to them safer.

## Things to be aware of

Mostly it's CopyOnWriteArrayList and ConcurrentHashMap with compute/IfAbsent.
I've tried hard to preserve even "unregister" order (sometimes it deletes the last instance, rather than the first one).
I've run this version several times with different number of threads (incl.11) for those 11 databases (tmpfs FTW).

## Things to worry about

There is a couple if (get() == null) put() usages of ConcurrentHashMap I've fixed, but there could be more.

nvoxland commented:
Thanks for working through that!

**Notes for reviewer**

Quick glance looked like thread-safe cleanup of code with no functionality or API changes. Removed some usage of SmartMap, did it remove all of it and we could remove that class? Or still some left and/or do we want to keep SmartMap around regardless (or at least deprecate)?

Questions I have:
- None

Potential risks:
- None

What could make the full review difficult:
- Nothing

ettavolt commented:
> Removed some usage of SmartMap, did it remove all of it and we could remove that class? Or still some left and/or do we want to keep SmartMap around regardless (or at least deprecate)?

SmartMap is still used by AbstractExtensibleObject. In a one private field, accessible through three methods. Relevant part may be inlined, I guess.

If you'd like, I can do this too. I enjoy erasing deprecated code. :smile:

nvoxland commented:
If you're good with inlining the usage on this PR @ettavolt I'd appreciate it :) 

Probably can't completely remove SmartMap yet, since it might be used by extensions. But you could mark it as deprecated for now at least.

ettavolt committed:
Safe concurrent updates of many un-Scoped factories state.
Less locking when using Scope.

Deprecate SmartMap.

MalloD12 committed:
Merge conflicts resolved - Merge branch 'concurrent-factories' of https://github.com/ettavolt/liquibase into ettavolt-concurrent-factories

MalloD12 committed:
Removed method accidentally added by mistake when resolving merge conflicts.

MalloD12 committed:
Merge remote-tracking branch 'origin/master' into ettavolt-concurrent-factories

MalloD12 committed:
Merge remote-tracking branch 'origin/master' into ettavolt-concurrent-factories

MalloD12 reviewed (commented):
@nvoxland , I have reviewed this PR and changes made here looks consistent to me across the different updated files. But from my current java/product knowledge I cannot guarantee there is not a potential risk. Build and tests have been executed successfully.

Could you please review this PR and approve it if looks good to you too?

Thanks,
Daniel.

MalloD12 commented:
Hi @ettavolt, after triggering our GH jobs I saw sonar is failing due to a few [bugs/code smells](https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&sinceLeakPeriod=true&pullRequest=3561&id=liquibase). Would you mind taking a look to it?

Thanks,
Daniel.

ettavolt commented:
Sure!

ettavolt commented:
Sorry, I've already messed history with my shallow-copy approach, before I realized that it would have been better to append rather than squash.

MalloD12 committed:
Merge remote-tracking branch 'origin/master' into ettavolt-concurrent-factories

# Conflicts:
#	liquibase-core/src/main/java/liquibase/Scope.java
#	liquibase-core/src/main/java/liquibase/serializer/ChangeLogSerializerFactory.java
#	liquibase-core/src/main/java/liquibase/serializer/SnapshotSerializerFactory.java
#	liquibase-core/src/main/java/liquibase/snapshot/SnapshotIdService.java
#	liquibase-core/src/main/java/liquibase/util/ObjectUtil.java

MalloD12 committed:
- Update BaseSerializerFactory to make serializers now be a ConcurrentHashMap instance, instead of just a HashMap.
- Remove serializers class attribute, since it's already defined in parent class.

MalloD12 committed:
Merge remote-tracking branch 'origin/master' into ettavolt-concurrent-factories

MalloD12 committed:
Removed unsed import.

MalloD12 commented:
Hi @ettavolt, 

I have resolved some merge conflicts, and tried to build again but am still seeing some sonar issues/code smells. Do you think you can fix them, so we can move forward with this PR?

Thanks,
Daniel.

ettavolt commented:
Yep!
(This time I'll try to make your reviewing job easier 😀)

ettavolt committed:
Rename some shadowing variables.
The copy of Scope method simplified and deduped.
Calm Sonar around double-check pattern.
Remove empty file - result of merge conflict.

MalloD12 committed:
Merge remote-tracking branch 'origin/master' into ettavolt-concurrent-factories

# Conflicts:
#	liquibase-core/src/main/java/liquibase/changelog/ChangeLogHistoryServiceFactory.java
#	liquibase-core/src/main/java/liquibase/diff/DiffGeneratorFactory.java
#	liquibase-core/src/main/java/liquibase/lockservice/LockServiceFactory.java
#	liquibase-core/src/main/java/liquibase/parser/SnapshotParserFactory.java
#	liquibase-core/src/main/java/liquibase/serializer/ChangeLogSerializerFactory.java
#	liquibase-core/src/main/java/liquibase/serializer/ReflectionSerializer.java
#	liquibase-core/src/main/java/liquibase/serializer/SnapshotSerializerFactory.java

MalloD12 committed:
Merge branch 'concurrent-factories' of https://github.com/ettavolt/liquibase into ettavolt-concurrent-factories

MalloD12 committed:
Merge remote-tracking branch 'origin/master' into ettavolt-concurrent-factories

# Conflicts:
#	liquibase-core/src/main/java/liquibase/serializer/ReflectionSerializer.java

MalloD12 commented on liquibase-core/src/main/java/liquibase/diff/output/StandardObjectChangeFilter.java:
```diff
@@ -36,8 +36,12 @@ public class StandardObjectChangeFilter implements ObjectChangeFilter {
     public StandardObjectChangeFilter(FilterType type, String filter) {
         this.filterType = type;
         if (databaseObjects.isEmpty()) {
```
@ettavolt is the first `IF` required because we have the same comparison inside of the `sinchronized` block?

ettavolt replied:
Rereading Wiki, I thought there could be a problem with nothing being `volatile`, potentially giving different results for the second conditional (which is inside `synchronized`). How, java.util.SynchornizedCollections don't use any `volatile` to provide thread-safety guarantee.

That said, current code relies on the fact that there is just one instance of writing, which might intendedly change in the future. To be on a safe side, let me initialize the field to null, and then set it to an unmodifiable list.

MalloD12 commented on liquibase-core/src/main/java/liquibase/statement/ExecutablePreparedStatementBase.java:
```diff
@@ -122,50 +102,21 @@ public void execute(PreparedStatementFactory factory) throws DatabaseException {
-                if (executeWithFlagsMethod == null) {
-                    executeWithFlagsMethod = stmt.getClass().getMethod("executeWithFlags", int.class);
-                    executeWithFlagsMethod.setAccessible(true);
+                if (postgresExecuteWithFlagsMethod == null) {
```
Same thing here, we have the same comparison outside and inside of the synchronized block.

MalloD12 replied:
@ettavolt don't we need to also make the same change you made on StandardObjectChangeFilter here too by making `executeWithFlagsMethod` a volatile field as well?

ettavolt replied:
`postgresExecuteWithFlagsMethod` is already defined as `volatile`. (It was earlier version that used `executeWithFlagsMethod` name).

MalloD12 replied:
You're totally right, I think I saw it wrong. Thank you!

ettavolt commented:
Yup, it's  https://en.m.wikipedia.org/wiki/Double-checked_locking

MalloD12 committed:
Merge branch 'master' into concurrent-factories

MalloD12 commented:
**Approved**.

## Review and testing results:

Besides, I am not a Thread safe expert, but I have seen code changes made and compared with what we had before and the changes look good to me. Some of the changes made have been to start using CopyOnWriteArrayList and ConcurrentHashMap data structures which are thread-safe and which allow us to use methods like `compute` and `IfAbsent`. As Nathan mentioned there are no code functionality or API changes made on this PR so I think we are ok to move forward. 

Build and tests passed successfully.

This is a good clean-up @ettavolt, thanks for taking care of this. And, to have fixed sonar issues/code smells.   

### Things to be aware of:
- None

### Things to worry about:
- Besides the testing, you have performed @ettavolt . I'm wondering whether we can add some unit tests around these changes. For example for BaseSerializerFactory where we have a couple of methods that are used in several places.

ettavolt commented:
As for testing BaseSerializerFactory: I think it's pretty nice covered by ChangeLogSerializerFactoryTest, which seem to call all four public methods. Just in case, I'm making BaseSerializerFactory package-local, so that it is not considered an API.

ettavolt committed:
Unmodifiable StandardObjectChangeFilter#databaseObjects.
Package-local BaseSerializerFactory to avoid it being considered an API.

ettavolt committed:
Mute Sonar's warning about StandardObjectChangeFilter#databaseObjects, explaining it partial immutability.

nvoxland reviewed (approved):
Nothing looks surprising or unexpected to me. Thanks!

ettavolt commented:
I guess that `com.datical.liquibase.ext.diff.FormattedDiffObject.getSerializableFields` should filter retrieved fields collection via stream(), or first create a new one off it.
It might even fix some bug!

ettavolt commented:
@XDelphiGrl, can you tell me from where is `com.datical.liquibase.ext.diff.FormattedDiffObject.getSerializableFields` getting the collection it tries to remove elements from? I can make that method to always return new modifiable collections.

filipelautert commented:
> @XDelphiGrl, can you tell me from where is `com.datical.liquibase.ext.diff.FormattedDiffObject.getSerializableFields` getting the collection it tries to remove elements from? I can make that method to always return new modifiable collections.

Hello @ettavolt - sure,  it gets from super:

```
    @Override
    public Set<String> getSerializableFields() {
        Set<String> fields = super.getSerializableFields();
        fields.remove("objectTypeHeader");
        return fields;
    }
```

and super is `AbstractLiquibaseSerializable `

ettavolt committed:
Merge remote-tracking branch 'origin/master' into concurrent-factories

# Conflicts:
#	liquibase-core/src/main/java/liquibase/changelog/ChangeLogHistoryServiceFactory.java
#	liquibase-standard/src/main/java/liquibase/AbstractExtensibleObject.java
#	liquibase-standard/src/main/java/liquibase/Scope.java
#	liquibase-standard/src/main/java/liquibase/change/ChangeFactory.java
#	liquibase-standard/src/main/java/liquibase/database/DatabaseFactory.java
#	liquibase-standard/src/main/java/liquibase/diff/DiffGeneratorFactory.java
#	liquibase-standard/src/main/java/liquibase/diff/output/changelog/ChangeGeneratorFactory.java
#	liquibase-standard/src/main/java/liquibase/diff/output/changelog/DiffToChangeLog.java
#	liquibase-standard/src/main/java/liquibase/lockservice/LockServiceFactory.java
#	liquibase-standard/src/main/java/liquibase/parser/ChangeLogParserFactory.java
#	liquibase-standard/src/main/java/liquibase/parser/NamespaceDetailsFactory.java
#	liquibase-standard/src/main/java/liquibase/parser/SnapshotParserFactory.java
#	liquibase-standard/src/main/java/liquibase/serializer/BaseSerializerFactory.java
#	liquibase-standard/src/main/java/liquibase/serializer/ChangeLogSerializerFactory.java
#	liquibase-standard/src/main/java/liquibase/serializer/ReflectionSerializer.java
#	liquibase-standard/src/main/java/liquibase/serializer/SnapshotSerializerFactory.java
#	liquibase-standard/src/main/java/liquibase/snapshot/SnapshotGeneratorFactory.java
#	liquibase-standard/src/main/java/liquibase/snapshot/SnapshotIdService.java
#	liquibase-standard/src/main/java/liquibase/sqlgenerator/SqlGeneratorComparator.java
#	liquibase-standard/src/main/java/liquibase/sqlgenerator/SqlGeneratorFactory.java

ettavolt committed:
Strictly-typed lambda-friendly PriorityArgsClosure in AbstractPluginFactory.

ettavolt committed:
Return modifiable set from AbstractLiquibaseSerializable#getSerializableFields for proprietary extension to consume.

ettavolt commented on liquibase-standard/src/main/java/liquibase/plugin/AbstractPluginFactory.java:
```diff
@@ -22,21 +24,18 @@ protected AbstractPluginFactory() {
      * @return null if no plugins are found or have a priority greater than zero.
      */
-    protected T getPlugin(final Object... args) {
+    protected T getPlugin(final PriorityArgsClosure<? super T> chooser) {
```
That's probably my third time using `super` type bound. ☺

MalloD12 committed:
Merge branch 'master' into concurrent-factories

MalloD12 commented:
Hi @ettavolt, 

Could you please look at the Postgres integration tests failing (https://github.com/liquibase/liquibase/actions/runs/5394370602/jobs/9795625506?pr=3561), and see whether the failures are related to your changes or not?

Thanks,
Daniel.

ettavolt commented:
Likely they are, because there are successful builds [before](https://github.com/liquibase/liquibase/actions/runs/5393962323) and [after](https://github.com/liquibase/liquibase/actions/runs/5401309105) that.

I'll try to run this integration test locally and see what it does wrong.

ettavolt committed:
JdbcConnection: don't return closed cached prepared statement.
Don't remove local uses of PreparedStatementFactory.

MalloD12 committed:
Merge branch 'master' into concurrent-factories

kevin-atx commented:
@ettavolt - any luck getting the integration tests to pass?

ettavolt commented:
I didn't see new tests results. I'll merge the mainline and fix the conflicts.

ettavolt committed:
Merge remote-tracking branch 'origin/master' into concurrent-factories

# Conflicts:
#	liquibase-standard/src/main/java/liquibase/change/ChangeFactory.java
#	liquibase-standard/src/main/java/liquibase/executor/jvm/JdbcExecutor.java
#	liquibase-standard/src/main/java/liquibase/parser/ChangeLogParserFactory.java
#	liquibase-standard/src/main/java/liquibase/plugin/AbstractPluginFactory.java

ettavolt committed:
Restore PrioritizedService inheriting Plugin.
Tests: reflect new way of PreparedStatementFactory caching statements.

ettavolt commented:
If you are referring to PostgreSQL integration tests: those were fixed like two months ago.

There were merge conflicts with master, I've resolved these now.
